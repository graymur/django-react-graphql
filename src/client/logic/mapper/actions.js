import { createAction } from 'redux-actions'

export const setConfig = createAction(
	'Mapper.setConfig',
	(config, savedMapping) => ({ config, savedMapping }),
)
export const setApps = createAction('Mapper.setApps', apps => apps)
export const addMapping = createAction(
	'Mapper.addMapping',
	(fromField, toField) => ({ fromField, toField }),
)

// export const resetState = createAction('Mapper.resetState', () => {})
// export const setWorkflowId = createAction(
// 	'Mapper.setWorkflowId',
// 	workflowId => workflowId,
// )
export const addDropZone = createAction('Mapper.addDropZone', dropzone => dropzone)
export const deleteMapping = createAction('Mapper.deleteMapping', field => field)
export const setMapping = createAction('Mapper.setMapping', mapping => mapping)
export const autoMapping = createAction('Mapper.autoMapping', () => {})
export const randomMapping = createAction('Mapper.randomMapping', () => {})
export const clearMapping = createAction('Mapper.clearMapping', () => {})
export const setActiveField = createAction('Mapper.setActiveField', field => field)
export const clearActiveField = createAction('Mapper.clearActiveField', () => {})
export const fieldDraggedAndDropped = createAction(
	'Mapper.fieldDraggedAndDropped',
	(field, eventX, eventY) => ({ field, eventX, eventY }),
)
