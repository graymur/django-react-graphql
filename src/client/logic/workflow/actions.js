import { createAction } from 'redux-actions'

export const setWorkflow = createAction(`Workflow.workflow`, workflow => workflow)

export const setWorkflowActiveState = createAction(
	`Workflow.activeState`,
	activeState => activeState,
)

export const setWorkflowCronParams = createAction(
	`Workflow.cronParams`,
	cronParams => cronParams,
)

export const addSettingsGroupWorkflow = createAction(
	`Workflow.group`,
	group => group,
)

export const addConnection = createAction(
	`Workflow.addConnection`,
	connection => connection,
)

export const addConnectionAndSetActive = createAction(
	`Workflow.addConnectionAndSetActive`,
	connection => connection,
)

export const setWidgetValue = createAction(
	`Workflow.setWidgetValue`,
	(name, value, validationErrors) => ({ name, value, validationErrors }),
)
