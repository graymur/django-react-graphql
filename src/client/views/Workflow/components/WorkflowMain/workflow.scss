@import "~style/lib";

.workflow {
    max-width: 80rem;
    margin: 0 auto;

    &._loading {
        min-height: 20rem;
    }
}