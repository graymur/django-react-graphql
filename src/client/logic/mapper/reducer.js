import * as actions from './actions'
import { handleActions } from 'redux-actions'
import findPotentialMapping from './utils/find-potential-mapping'
import randomMapping from './utils/random-mapping'
import autoMapping from './utils/auto-mapping'
import createMappingObject from './utils/create-mapping-object'
import { orderFromFields, orderToFields } from './utils/order-fields'
// import isString from 'lodash/isString'

export const MAPPER_REDUCER_KEY = 'Mapper'

// const dv = (title, v) => {
// 	if (!isString(title)) {
// 		console.log(JSON.stringify(title, null, 4))
// 	} else {
// 		console.log(title, JSON.stringify(v, null, 4))
// 	}
// }

export const initialState = {
	config: null,
	apps: [],
	fromFields: [],
	toFields: [],
	activeFields: undefined,
	potentialMappings: [],
	mapping: [],
	dropzones: [],
}

const setMappingInState = (state, mapping) => {
	const fromFields = orderFromFields(state.apps[0].fields, mapping).map(field => ({
		...field,
		mappingObject: mapping.find(x => x.from === field._id),
	}))
	const toFields = orderToFields(state.apps[1].fields, fromFields, mapping)

	return {
		...state,
		mapping,
		fromFields,
		toFields,
	}
}

const setApps = (state, { payload: apps }) => {
	const appsWithType = apps.map((app, index) => ({
		...app,
		// add __type field to each field to distinguish from from to
		fields: app.fields.map(x => ({
			...x,
			__type: index === 0 ? 'from' : 'to',
			appName: app.title,
		})),
	}))

	return {
		...state,
		apps: appsWithType,
		fromFields: appsWithType[0].fields,
		toFields: appsWithType[1].fields,
	}
}

const addMapping = (state, { payload: { fromField, toField } }) => {
	const from = fromField.__type === 'from' ? fromField : toField
	const to = fromField.__type === 'from' ? toField : fromField

	const mapping = [...state.mapping, createMappingObject(from, to)]

	return setMappingInState(state, mapping)
}

const setActiveField = (state, { payload: activeField }) => {
	const targets = activeField.__type === 'from' ? state.toFields : state.fromFields
	const potentialMappings = findPotentialMapping(
		activeField,
		targets.filter(x => !x.isMapped),
	).map(x => x._id)

	return {
		...state,
		potentialMappings,
		activeField,
	}
}

const fieldDraggedAndDropped = (state, { payload: { field, eventX, eventY } }) => {
	const dropzone = state.dropzones.find(
		({ rect }) =>
			eventX >= rect.left &&
			eventX <= rect.left + rect.width &&
			eventY >= rect.top &&
			eventY <= rect.top + rect.height,
	)

	return dropzone
		? addMapping(state, {
				payload: { fromField: field, toField: dropzone.field },
		  })
		: state
}

const setConfig = (state, { payload: { config, savedMapping } }) => {
	let newState = state
	newState = setApps(newState, { payload: config.apps })
	newState = setMappingInState(newState, savedMapping)
	return {
		...newState,
		config,
	}
}

export default handleActions(
	{
		[actions.setConfig]: setConfig,
		[actions.setApps]: setApps,
		// [actions.setWorkflowId]: (state, { payload }) => ({
		// 	...state,
		// 	workflowId: payload,
		// }),
		// [actions.resetState]: () => initialState,
		[actions.addMapping]: addMapping,
		[actions.setMapping]: (state, { payload: mapping }) =>
			setMappingInState(state, mapping),
		[actions.autoMapping]: state =>
			setMappingInState(
				state,
				autoMapping(state.apps[0].fields, state.apps[1].fields, {
					matchByFieldsNames: true,
				}),
			),
		[actions.clearMapping]: state => setMappingInState(state, []),
		[actions.clearActiveField]: state => {
			return {
				...state,
				activeField: undefined,
				dropzones: [],
				potentialMappings: [],
			}
		},
		[actions.addDropZone]: (state, { payload: dropzone }) => ({
			...state,
			dropzones: [...state.dropzones, dropzone],
		}),
		[actions.deleteMapping]: (state, { payload: field }) =>
			setMappingInState(
				state,
				state.mapping.filter(x => x.from !== field._id && x.to !== field._id),
			),
		[actions.setActiveField]: setActiveField,
		[actions.fieldDraggedAndDropped]: fieldDraggedAndDropped,
	},
	initialState,
)
