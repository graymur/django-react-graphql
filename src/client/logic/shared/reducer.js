import * as actions from './actions'
import { handleActions } from 'redux-actions'

export const SHARED_REDUCER_KEY = 'Shared'

export const initialState = {
	error: undefined,
	loading: false,
	isLoggedIn: true,
	backendError: null,
}

export default handleActions(
	{
		[actions.setLoading]: (state, { payload: loading }) => ({
			...state,
			loading,
		}),
		[actions.setError]: (state, { payload: error }) => ({
			...state,
			error,
		}),
		[actions.clearError]: state => ({
			...state,
			error: undefined,
		}),
		[actions.setIsLoggedIn]: (state, { payload: isLoggedIn }) => ({
			...state,
			isLoggedIn,
		}),
		[actions.setBackendError]: (state, { payload: backendError }) => ({
			...state,
			backendError,
		}),
		'@@router/LOCATION_CHANGE': state => ({
			...state,
			error: undefined,
		}),
	},
	initialState,
)
