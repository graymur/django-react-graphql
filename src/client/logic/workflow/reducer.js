import * as actions from './actions'
import { handleActions } from 'redux-actions'
import validateWorkflow from './utils/validate-workflow'
import omit from 'lodash/omit'

import getWidgetsValuesFromWorkflow from './utils/get-widgets-values-from-workflow'
import validateWidgets from './utils/validate-widgets'

export const WORKFLOW_REDUCER_KEY = 'Workflow'

export const initialState = {
	workflow: undefined,
	validationErrors: false,
}

/**
 * GraphQL API will return all objects fetched from Mongo with "_id" field, but webapp's API
 * replaces it with "id" field. We need to changed it to GraphQL variant maintain consistency
 * between data coming from GraphQL and webapp's REST endpoints
 * @param obj
 * @returns {{_id}}
 */
const replaceIdWithUnderscoredId = obj =>
	obj.id
		? {
				...omit(obj, 'id'),
				_id: obj.id,
		  }
		: obj

const addConnection = (state, { payload }) => ({
	...state,
	workflow: {
		...state.workflow,
		connectionObjects: [
			...state.workflow.connectionObjects,
			replaceIdWithUnderscoredId(payload),
		],
	},
	validationErrors: validateWorkflow(state.workflow),
})

export default handleActions(
	{
		[actions.setWorkflow]: (state, { payload: workflow }) => ({
			...state,
			initialWidgetsValues: getWidgetsValuesFromWorkflow(workflow),
			widgetsValues: {
				...getWidgetsValuesFromWorkflow(workflow),
				...state.widgetsValues,
			},
			widgetsErrors: validateWidgets(workflow),
			workflow,
			validationErrors: validateWorkflow(workflow),
		}),
		[actions.setWidgetValue]: (
			state,
			{ payload: { name, value, validationErrors } },
		) => {
			return {
				...state,
				widgetsValues: {
					...state.widgetsValues,
					[name]: value,
				},
				widgetsErrors: {
					...state.widgetsErrors,
					[name]: validationErrors,
				},
			}
		},
		[actions.setWorkflowActiveState]: (state, { payload }) => ({
			...state,
			workflow: { ...state.workflow, ...payload },
			validationErrors: validateWorkflow(state.workflow),
		}),
		[actions.setWorkflowCronParams]: (state, { payload }) => ({
			...state,
			workflow: { ...state.workflow, cron_params: payload },
			validationErrors: validateWorkflow(state.workflow),
		}),
		[actions.addConnection]: addConnection,
		/**
		 * Add new connection object to list of existing connection
		 * objects and set it as active in workflow's config
		 * @param state
		 * @param payload
		 * @returns {{workflow: {config: {}}}}
		 */
		[actions.addConnectionAndSetActive]: (state, { payload }) => {
			const newState = addConnection(state, { payload })

			return {
				...newState,
				workflow: {
					...newState.workflow,
					config: {
						...newState.workflow.config,
						[payload.name]: {
							...newState.workflow.config[payload.name],
							connect_id: payload.id,
						},
					},
				},
			}
		},
	},
	initialState,
)
