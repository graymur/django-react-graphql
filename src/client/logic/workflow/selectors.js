import { createSelector } from 'reselect'
import { WORKFLOW_REDUCER_KEY } from './reducer'

export const selectState = state => state[WORKFLOW_REDUCER_KEY]
export const workflow = createSelector(selectState, main => main.workflow)
export const validationErrors = createSelector(
	selectState,
	main => main.validationErrors,
)
export const initialWidgetsValues = createSelector(
	selectState,
	main => main.initialWidgetsValues,
)
export const widgetsValues = createSelector(selectState, main => main.widgetsValues)
export const widgetsErrors = createSelector(selectState, main => main.widgetsErrors)
