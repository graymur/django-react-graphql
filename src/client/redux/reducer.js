import { combineReducers } from 'redux'
import { routerReducer } from 'react-router-redux'
import shared, { SHARED_REDUCER_KEY } from 'client/logic/shared/reducer'
import mapper, { MAPPER_REDUCER_KEY } from 'client/logic/mapper/reducer'
import workflow, { WORKFLOW_REDUCER_KEY } from 'client/logic/workflow/reducer'

export default combineReducers({
	[SHARED_REDUCER_KEY]: shared,
	[MAPPER_REDUCER_KEY]: mapper,
	[WORKFLOW_REDUCER_KEY]: workflow,
	routing: routerReducer,
})
