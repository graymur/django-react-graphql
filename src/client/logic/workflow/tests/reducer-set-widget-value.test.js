import { setWorkflow, setWidgetValue } from '../actions'
import reducer, { initialState } from '../reducer'
import workflow from './data/workflow'

test('Sets valid widget value in state', () => {
	let state = reducer(initialState, setWorkflow(workflow))
	state = reducer(state, setWidgetValue('listId', 1, false))
	expect(state).toMatchSnapshot()
})

test('Sets invalid widget value in state', () => {
	let state = reducer(initialState, setWorkflow(workflow))
	state = reducer(
		state,
		setWidgetValue('listId', false, {
			code: 1,
			messageWidgetId: 'list_id-error',
		}),
	)

	expect(state).toMatchSnapshot()
})
