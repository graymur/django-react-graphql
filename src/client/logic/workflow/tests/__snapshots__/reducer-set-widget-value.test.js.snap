// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sets invalid widget value in state 1`] = `
Object {
  "initialWidgetsValues": Object {
    "listId": undefined,
    "trigger": "get_subscribers",
    "visual-filter": undefined,
    "visual-mapping": undefined,
  },
  "validationErrors": undefined,
  "widgetsErrors": Object {
    "listId": Object {
      "code": 1,
      "messageWidgetId": "list_id-error",
    },
    "trigger": false,
    "visual-filter": undefined,
    "visual-mapping": Object {
      "code": 1,
      "messageWidgetId": "",
    },
  },
  "widgetsValues": Object {
    "listId": false,
    "trigger": "get_subscribers",
    "visual-filter": undefined,
    "visual-mapping": undefined,
  },
  "workflow": Object {
    "__typename": "Workflow",
    "_id": "5b6c4fb9b6e34d0bb03a3ccb",
    "about": "",
    "apps": Array [
      Object {
        "__typename": "App",
        "full_name": "MailChimp",
        "name": "cn_mailchimp",
      },
      Object {
        "__typename": "App",
        "full_name": "Dotloop",
        "name": "cn_dotloop",
      },
    ],
    "config": Object {
      "cn_dotloop": Object {
        "connect_id": "5b69ab94c409d8000439d835",
        "full_name_app": "Dotloop",
      },
      "cn_mailchimp": Object {
        "connect_id": "5b69a78ec409d8000439d833",
        "saved_params": Object {
          "listId": "29aaf507bc",
          "trigger": "get_subscribers",
        },
      },
      "global": Object {
        "access_key": null,
        "email": "sergey@apination.com",
        "trigger_id": "5b6c4fb9b6e34d0bb03a3cca",
        "user_id": 422,
      },
    },
    "configuration": Array [
      Object {
        "app": "cn_mailchimp",
        "label": "Trigger",
        "title": "Create or update contact",
        "validations": Array [
          Object {
            "message": "#list_id-error",
            "type": "required",
            "with": Array [
              "mailing_list_id",
            ],
          },
        ],
        "widgets": Array [
          Object {
            "class": "hidden",
            "name": "trigger",
            "node": "input",
            "type": "text",
            "value": "get_subscribers",
            "wrapper-class": "hidden",
          },
          Object {
            "node": "header",
            "text": "Mailing list",
          },
          Object {
            "class": "workflow_select_mailing",
            "data-method": "cn_mailchimp.get_mailing_lists",
            "depend": "mapping",
            "id": "mailing_list_id",
            "method": Object {
              "map": Object {
                "text": "name",
                "value": "id",
              },
              "method": "cn_mailchimp.get_mailing_lists",
            },
            "name": "listId",
            "node": "select-with-refresh",
            "raw-options": "<option selected disabled value=\\"CASE_NEED_SELECT\\" color:\\"#858585;\\">Select list</option>",
          },
          Object {
            "id": "list_id-error",
            "node": "validation-message",
            "text": "Select mailing list",
          },
          Object {
            "node": "save-block",
          },
        ],
      },
      Object {
        "app": "cn_dotloop",
        "label": "Action",
        "title": "Create or update person",
        "validations": Array [],
        "widgets": Array [],
      },
      Object {
        "app": "cn_mailchimp",
        "groupClassName": "filter-widget",
        "label": "Filter",
        "title": "Filter contacts",
        "widgets": Array [
          Object {
            "config": Object {
              "appName": "cn_mailchimp",
              "entityName": "subscriber",
              "fields": Object {
                "method": "cn_mailchimp.get_subscriber_config_trigger",
                "params": Array [
                  "listId",
                ],
              },
            },
            "id": "visual-filter",
            "name": "visual-filter",
            "node": "visual-filter",
          },
          Object {
            "node": "save-block",
          },
        ],
      },
      Object {
        "app": "cn_mailchimp",
        "groupClassName": "mapping-widget",
        "label": "Mapping",
        "title": "MailChimp subscribers to Dotloop people",
        "validations": Array [
          Object {
            "field": "visual-mapping",
            "type": "required",
          },
        ],
        "widgets": Array [
          Object {
            "config": Object {
              "cn_dotloop": Object {
                "appName": "cn_dotloop",
                "entityName": "person",
                "fields": Array [
                  Object {
                    "_id": "dotloopFirstName",
                    "autoMapping": Object {
                      "name": "firstName",
                    },
                    "sample": "John",
                    "setter": "firstName",
                    "title": "First name",
                  },
                  Object {
                    "_id": "dotloopRole",
                    "sample": "BUYER",
                    "setter": "role",
                    "title": "Role",
                  },
                ],
              },
              "cn_mailchimp": Object {
                "appName": "cn_mailchimp",
                "entityName": "subscriber",
                "fields": Object {
                  "method": "cn_mailchimp.get_subscriber_config_trigger",
                  "params": Array [
                    "listId",
                  ],
                },
              },
            },
            "id": "visual-mapping",
            "name": "visual-mapping",
            "node": "visual-mapping",
          },
          Object {
            "node": "save-block",
          },
        ],
      },
    ],
    "connectionObjects": Array [],
    "connections": Array [
      "cn_mailchimp",
      "cn_dotloop",
    ],
    "cron_params": Object {
      "minutes": 10,
    },
    "full_description": "",
    "is_enabled": false,
    "name": "mailchimp_subscribers_to_dotloop_people",
    "saved_params": Object {
      "cn_mailchimp": Object {
        "trigger": "get_subscribers",
      },
    },
    "short_description": "",
    "status": "unused",
    "workflow": Array [
      Object {
        "cn_mailchimp_get_updated_subscribers": Object {
          "apiKey": "$config.cn_mailchimp.params.apiKey",
          "listId": "$config.cn_mailchimp.saved_params.listId",
          "mainLogId": "$config.runtime.saga_id",
        },
      },
      Object {
        "cn_dotloop_update_contacts": Object {
          "mainLogId": "$config.runtime.saga_id",
          "token": "$config.cn_dotloop.params.token",
        },
      },
    ],
  },
}
`;

exports[`Sets valid widget value in state 1`] = `
Object {
  "initialWidgetsValues": Object {
    "listId": undefined,
    "trigger": "get_subscribers",
    "visual-filter": undefined,
    "visual-mapping": undefined,
  },
  "validationErrors": undefined,
  "widgetsErrors": Object {
    "listId": false,
    "trigger": false,
    "visual-filter": undefined,
    "visual-mapping": Object {
      "code": 1,
      "messageWidgetId": "",
    },
  },
  "widgetsValues": Object {
    "listId": 1,
    "trigger": "get_subscribers",
    "visual-filter": undefined,
    "visual-mapping": undefined,
  },
  "workflow": Object {
    "__typename": "Workflow",
    "_id": "5b6c4fb9b6e34d0bb03a3ccb",
    "about": "",
    "apps": Array [
      Object {
        "__typename": "App",
        "full_name": "MailChimp",
        "name": "cn_mailchimp",
      },
      Object {
        "__typename": "App",
        "full_name": "Dotloop",
        "name": "cn_dotloop",
      },
    ],
    "config": Object {
      "cn_dotloop": Object {
        "connect_id": "5b69ab94c409d8000439d835",
        "full_name_app": "Dotloop",
      },
      "cn_mailchimp": Object {
        "connect_id": "5b69a78ec409d8000439d833",
        "saved_params": Object {
          "listId": "29aaf507bc",
          "trigger": "get_subscribers",
        },
      },
      "global": Object {
        "access_key": null,
        "email": "sergey@apination.com",
        "trigger_id": "5b6c4fb9b6e34d0bb03a3cca",
        "user_id": 422,
      },
    },
    "configuration": Array [
      Object {
        "app": "cn_mailchimp",
        "label": "Trigger",
        "title": "Create or update contact",
        "validations": Array [
          Object {
            "message": "#list_id-error",
            "type": "required",
            "with": Array [
              "mailing_list_id",
            ],
          },
        ],
        "widgets": Array [
          Object {
            "class": "hidden",
            "name": "trigger",
            "node": "input",
            "type": "text",
            "value": "get_subscribers",
            "wrapper-class": "hidden",
          },
          Object {
            "node": "header",
            "text": "Mailing list",
          },
          Object {
            "class": "workflow_select_mailing",
            "data-method": "cn_mailchimp.get_mailing_lists",
            "depend": "mapping",
            "id": "mailing_list_id",
            "method": Object {
              "map": Object {
                "text": "name",
                "value": "id",
              },
              "method": "cn_mailchimp.get_mailing_lists",
            },
            "name": "listId",
            "node": "select-with-refresh",
            "raw-options": "<option selected disabled value=\\"CASE_NEED_SELECT\\" color:\\"#858585;\\">Select list</option>",
          },
          Object {
            "id": "list_id-error",
            "node": "validation-message",
            "text": "Select mailing list",
          },
          Object {
            "node": "save-block",
          },
        ],
      },
      Object {
        "app": "cn_dotloop",
        "label": "Action",
        "title": "Create or update person",
        "validations": Array [],
        "widgets": Array [],
      },
      Object {
        "app": "cn_mailchimp",
        "groupClassName": "filter-widget",
        "label": "Filter",
        "title": "Filter contacts",
        "widgets": Array [
          Object {
            "config": Object {
              "appName": "cn_mailchimp",
              "entityName": "subscriber",
              "fields": Object {
                "method": "cn_mailchimp.get_subscriber_config_trigger",
                "params": Array [
                  "listId",
                ],
              },
            },
            "id": "visual-filter",
            "name": "visual-filter",
            "node": "visual-filter",
          },
          Object {
            "node": "save-block",
          },
        ],
      },
      Object {
        "app": "cn_mailchimp",
        "groupClassName": "mapping-widget",
        "label": "Mapping",
        "title": "MailChimp subscribers to Dotloop people",
        "validations": Array [
          Object {
            "field": "visual-mapping",
            "type": "required",
          },
        ],
        "widgets": Array [
          Object {
            "config": Object {
              "cn_dotloop": Object {
                "appName": "cn_dotloop",
                "entityName": "person",
                "fields": Array [
                  Object {
                    "_id": "dotloopFirstName",
                    "autoMapping": Object {
                      "name": "firstName",
                    },
                    "sample": "John",
                    "setter": "firstName",
                    "title": "First name",
                  },
                  Object {
                    "_id": "dotloopRole",
                    "sample": "BUYER",
                    "setter": "role",
                    "title": "Role",
                  },
                ],
              },
              "cn_mailchimp": Object {
                "appName": "cn_mailchimp",
                "entityName": "subscriber",
                "fields": Object {
                  "method": "cn_mailchimp.get_subscriber_config_trigger",
                  "params": Array [
                    "listId",
                  ],
                },
              },
            },
            "id": "visual-mapping",
            "name": "visual-mapping",
            "node": "visual-mapping",
          },
          Object {
            "node": "save-block",
          },
        ],
      },
    ],
    "connectionObjects": Array [],
    "connections": Array [
      "cn_mailchimp",
      "cn_dotloop",
    ],
    "cron_params": Object {
      "minutes": 10,
    },
    "full_description": "",
    "is_enabled": false,
    "name": "mailchimp_subscribers_to_dotloop_people",
    "saved_params": Object {
      "cn_mailchimp": Object {
        "trigger": "get_subscribers",
      },
    },
    "short_description": "",
    "status": "unused",
    "workflow": Array [
      Object {
        "cn_mailchimp_get_updated_subscribers": Object {
          "apiKey": "$config.cn_mailchimp.params.apiKey",
          "listId": "$config.cn_mailchimp.saved_params.listId",
          "mainLogId": "$config.runtime.saga_id",
        },
      },
      Object {
        "cn_dotloop_update_contacts": Object {
          "mainLogId": "$config.runtime.saga_id",
          "token": "$config.cn_dotloop.params.token",
        },
      },
    ],
  },
}
`;
